File: sdk/erlang/src/next_loggers.erl
-module(next_loggers).

-export([
    new/3,
    new/4,
    current_context/0,
    with_context/2,
    event/4,
    send/1,
    with_otel/2,
    use_otel/1,
    not_otel/1,
    reset_otel/1,
    is_otel_enabled/2,
    set_otel_enabled/2,
    log/4,
    info/3,
    error/3,
    otel_transport/1,
    supabase_transport/1
]).

-define(SCHEMA, <<"next-loggers/v1">>).
-define(CONTEXT_KEY, '$next_loggers_context').

new(AppName, Runtime, Transports) ->
    new(AppName, Runtime, #{}, Transports).

new(AppName, Runtime, Fields, Transports)
        when is_binary(AppName), is_binary(Runtime), is_map(Fields), is_list(Transports) ->
    case byte_size(AppName) of
        0 -> error({invalid_app_name, AppName});
        _ -> #{app_name => AppName, runtime => Runtime, fields => Fields, transports => Transports, otel => true}
    end.

current_context() ->
    case erlang:get(?CONTEXT_KEY) of
        undefined -> #{};
        Context when is_map(Context) -> Context
    end.

%% Process dictionary scope is local to the current lightweight BEAM process.
%% The previous frame is restored even when the callback throws.
with_context(Context, Fun) when is_map(Context), is_function(Fun, 0) ->
    Previous = erlang:get(?CONTEXT_KEY),
    erlang:put(?CONTEXT_KEY, Context),
    try Fun()
    after
        case Previous of
            undefined -> erlang:erase(?CONTEXT_KEY);
            _ -> erlang:put(?CONTEXT_KEY, Previous)
        end
    end.

info(Logger, Message, Fields) ->
    log(Logger, <<"INFO">>, Message, Fields).

error(Logger, Message, Fields) ->
    log(Logger, <<"ERROR">>, Message, Fields).

log(Logger, Level, Message, EventFields)
        when is_map(Logger), is_binary(Level), is_binary(Message), is_map(EventFields) ->
    send(event(Logger, Level, Message, EventFields)).

event(Logger, Level, Message, EventFields)
        when is_map(Logger), is_binary(Level), is_binary(Message), is_map(EventFields) ->
    #{
        kind => event,
        logger => Logger,
        level => Level,
        message => Message,
        fields => EventFields,
        context => current_context(),
        otel_enabled => undefined
    }.

set_otel_enabled(Logger, Enabled) when is_map(Logger), is_boolean(Enabled) ->
    maps:put(otel, Enabled, Logger).

use_otel(Value) when is_map(Value) ->
    case maps:get(kind, Value, logger) of
        event -> with_otel(Value, true);
        logger -> set_otel_enabled(Value, true)
    end.

not_otel(Value) when is_map(Value) ->
    case maps:get(kind, Value, logger) of
        event -> with_otel(Value, false);
        logger -> set_otel_enabled(Value, false)
    end.

with_otel(Event, Enabled) when is_map(Event), is_boolean(Enabled) ->
    maps:put(otel_enabled, Enabled, Event).

reset_otel(Event) when is_map(Event) ->
    maps:put(otel_enabled, undefined, Event).

is_otel_enabled(Event, Fallback) when is_map(Event), is_boolean(Fallback) ->
    case maps:get(otel_enabled, Event, undefined) of
        undefined -> Fallback;
        Enabled -> Enabled
    end.

send(Event) when is_map(Event) ->
    Logger = maps:get(logger, Event),
    Level = maps:get(level, Event),
    Message = maps:get(message, Event),
    EventFields = maps:get(fields, Event),
    Context = maps:get(context, Event, #{}),
    LoggerFields = maps:get(fields, Logger, #{}),
    ContextFields = maps:get(fields, Context, #{}),
    Fields0 = maps:merge(LoggerFields, ContextFields),
    Fields1 = put_optional(<<"otel.span_id">>, maps:get(span_id, Context, undefined), Fields0),
    Fields2 = maps:put(<<"otel.trace_flags">>, maps:get(trace_flags, Context, 0), Fields1),
    Fields3 = put_optional(<<"otel.trace_state">>, maps:get(trace_state, Context, undefined), Fields2),
    Fields = maps:merge(Fields3, EventFields),
    TraceId = maps:get(trace_id, Context, undefined),
    Record0 = #{
        schema => ?SCHEMA,
        id => unique_id(),
        timestamp => timestamp(),
        level => Level,
        runtime => maps:get(runtime, Logger),
        appName => maps:get(app_name, Logger),
        message => Message,
        values => [Message],
        fields => Fields
    },
    Record1 = put_optional(traceId, TraceId, Record0),
    Record2 = case TraceId of
        undefined -> Record1;
        <<>> -> Record1;
        _ -> maps:put(traceIds, [TraceId], Record1)
    end,
    Tags = maps:get(tags, Context, []),
    Record = case Tags of
        [] -> Record2;
        _ -> maps:put(tags, Tags, Record2)
    end,
    OtelEnabled = is_otel_enabled(Event, maps:get(otel, Logger, true)),
    lists:foreach(
        fun(Transport) ->
            case is_otel_transport(Transport) andalso not OtelEnabled of
                true -> ok;
                false -> ok = deliver_transport(Transport, Record)
            end
        end,
        maps:get(transports, Logger, [])
    ),
    Record.

%% Application-owned OpenTelemetry adapter. It emits data but never installs
%% a tracer, logger provider, context manager, or automatic instrumentation.
otel_transport(Sink) when is_function(Sink, 1) ->
    {opentelemetry, fun(Record) ->
        Level = maps:get(level, Record),
        Fields = maps:get(fields, Record, #{}),
        Attributes0 = #{
            <<"service.name">> => maps:get(appName, Record),
            <<"next_logger.schema">> => maps:get(schema, Record),
            <<"next_logger.runtime">> => maps:get(runtime, Record),
            <<"log.record.uid">> => maps:get(id, Record)
        },
        Attributes1 = maps:fold(
            fun(Key, Value, Acc) ->
                maps:put(<<"next_logger.field.", (key_binary(Key))/binary>>, Value, Acc)
            end,
            Attributes0,
            Fields
        ),
        Attributes = put_optional(<<"trace.id">>, maps:get(traceId, Record, undefined), Attributes1),
        Sink(#{
            body => maps:get(message, Record),
            severityText => Level,
            severityNumber => severity_number(Level),
            timestamp => maps:get(timestamp, Record),
            attributes => Attributes
        }),
        ok
    end}.

%% Client transport with an injected authenticated Supabase sender.
supabase_transport(Sender) when is_function(Sender, 1) ->
    {supabase, fun(Record) -> Sender(Record), ok end}.

is_otel_transport({Name, _}) when Name =:= opentelemetry; Name =:= <<"opentelemetry">> -> true;
is_otel_transport(_) -> false.

deliver_transport({_Name, Transport}, Record) when is_function(Transport, 1) -> Transport(Record);
deliver_transport(Transport, Record) when is_function(Transport, 1) -> Transport(Record).

severity_number(<<"TRACE">>) -> 1;
severity_number(<<"DEBUG">>) -> 5;
severity_number(<<"INFO">>) -> 9;
severity_number(<<"WARN">>) -> 13;
severity_number(<<"ERROR">>) -> 17;
severity_number(<<"FATAL">>) -> 21.

put_optional(_Key, undefined, Map) -> Map;
put_optional(_Key, <<>>, Map) -> Map;
put_optional(Key, Value, Map) -> maps:put(Key, Value, Map).

key_binary(Key) when is_binary(Key) -> Key;
key_binary(Key) when is_atom(Key) -> atom_to_binary(Key, utf8);
key_binary(Key) -> unicode:characters_to_binary(io_lib:format("~p", [Key])).

unique_id() ->
    integer_to_binary(erlang:unique_integer([monotonic, positive])).

timestamp() ->
    list_to_binary(calendar:system_time_to_rfc3339(erlang:system_time(second), [
        {unit, second},
        {offset, "Z"}
    ])).

File: sdk/elixir/lib/next_loggers.ex
defmodule ORESoftware.NextLoggers do
  @moduledoc """
  Dependency-free Elixir implementation of the `next-loggers/v1` contract.

  Context is scoped to the current BEAM process and restored after the callback.
  OpenTelemetry and Supabase are injected transports; this module never installs
  global providers, telemetry handlers, or automatic instrumentation.
  """

  @schema "next-loggers/v1"
  @context_key {__MODULE__, :context}

  def schema, do: @schema

  def new(app_name, opts \\ []) when is_binary(app_name) and is_list(opts) do
    if String.trim(app_name) == "", do: raise(ArgumentError, "app_name must not be empty")

    %{
      app_name: app_name,
      name: Keyword.get(opts, :name),
      runtime: Keyword.get(opts, :runtime, "elixir"),
      fields: Map.new(Keyword.get(opts, :fields, %{})),
      transports: List.wrap(Keyword.get(opts, :transports, [])),
      # Default routing for OTEL transports when a call makes no choice.
      otel: Keyword.get(opts, :otel, true),
      id_factory: Keyword.get(opts, :id_factory, &default_id/0),
      clock: Keyword.get(opts, :clock, &default_clock/0)
    }
  end

  def current_context do
    Process.get(@context_key, %{})
  end

  def with_context(context, callback) when is_map(context) and is_function(callback, 0) do
    previous = Process.get(@context_key, :__missing__)
    Process.put(@context_key, context)

    try do
      callback.()
    after
      case previous do
        :__missing__ -> Process.delete(@context_key)
        value -> Process.put(@context_key, value)
      end
    end
  end

  @doc "Derived logger that delivers every record to OTEL transports (the default)."
  def use_otel(logger) when is_map(logger), do: with_otel(logger, true)

  @doc """
  Derived logger that keeps records off OTEL transports; every other transport
  still receives them: `logger |> ORESoftware.NextLoggers.not_otel() |> info("msg")`.
  """
  def not_otel(logger) when is_map(logger), do: with_otel(logger, false)

  def with_otel(logger, enabled) when is_map(logger) and is_boolean(enabled),
    do: Map.put(logger, :otel, enabled)

  def otel_enabled(logger) when is_map(logger), do: Map.get(logger, :otel, true)

  def info(logger, message, fields \\ %{}), do: log(logger, "INFO", message, fields)
  def warn(logger, message, fields \\ %{}), do: log(logger, "WARN", message, fields)
  def error(logger, message, fields \\ %{}), do: log(logger, "ERROR", message, fields)

  def log(logger, level, message, event_fields)
      when is_map(logger) and is_binary(level) and is_binary(message) and is_map(event_fields) do
    log(logger, level, message, event_fields, otel_enabled(logger))
  end

  @doc """
  `otel` overrides this call's routing: `true` forces delivery to OTEL
  transports, `false` skips them.
  """
  def log(logger, level, message, event_fields, otel)
      when is_map(logger) and is_binary(level) and is_binary(message) and is_map(event_fields) and
             is_boolean(otel) do
    context = current_context()

    fields =
      logger.fields
      |> Map.merge(Map.get(context, :fields, %{}))
      |> put_optional("otel.span_id", Map.get(context, :span_id))
      |> Map.put("otel.trace_flags", Map.get(context, :trace_flags, 0))
      |> put_optional("otel.trace_state", Map.get(context, :trace_state))
      |> Map.merge(event_fields)

    trace_id = Map.get(context, :trace_id)

    record =
      %{
        "schema" => @schema,
        "id" => logger.id_factory.(),
        "timestamp" => logger.clock.(),
        "level" => level,
        "runtime" => logger.runtime,
        "appName" => logger.app_name,
        "message" => message,
        "values" => [message],
        "fields" => fields
      }
      |> put_optional("name", logger.name)
      |> put_optional("traceId", trace_id)
      |> maybe_put_trace_ids(trace_id)
      |> maybe_put_tags(Map.get(context, :tags, []))

    Enum.each(logger.transports, &deliver(&1, record, otel))

    record
  end

  # otel_transport/1 returns a tagged transport so routing can skip it without
  # inspecting the closure.
  defp deliver({:otel, _sink}, _record, false), do: :ok
  defp deliver({:otel, sink}, record, _otel) when is_function(sink, 1), do: call(sink, record)
  defp deliver(sink, record, _otel) when is_function(sink, 1), do: call(sink, record)

  defp call(sink, record) do
    case sink.(record) do
      :ok -> :ok
      other -> raise "transport returned #{inspect(other)}"
    end
  end

  def otel_transport(sink) when is_function(sink, 1) do
    {:otel,
     fn record ->
       attributes =
         %{
           "service.name" => record["appName"],
           "next_logger.schema" => record["schema"],
           "next_logger.runtime" => record["runtime"],
           "log.record.uid" => record["id"]
         }
         |> put_optional("trace.id", record["traceId"])
         |> Map.merge(
           Map.new(record["fields"], fn {key, value} ->
             {"next_logger.field.#{key}", value}
           end)
         )

       sink.(%{
         "body" => record["message"],
         "severityText" => record["level"],
         "severityNumber" => severity_number(record["level"]),
         "timestamp" => record["timestamp"],
         "attributes" => attributes
       })

       :ok
     end}
  end

  def supabase_transport(sender) when is_function(sender, 1) do
    fn record ->
      sender.(record)
      :ok
    end
  end

  defp maybe_put_trace_ids(map, nil), do: map
  defp maybe_put_trace_ids(map, ""), do: map
  defp maybe_put_trace_ids(map, trace_id), do: Map.put(map, "traceIds", [trace_id])

  defp maybe_put_tags(map, []), do: map
  defp maybe_put_tags(map, tags), do: Map.put(map, "tags", Enum.uniq(tags))

  defp put_optional(map, _key, nil), do: map
  defp put_optional(map, _key, ""), do: map
  defp put_optional(map, key, value), do: Map.put(map, key, value)

  defp severity_number("TRACE"), do: 1
  defp severity_number("DEBUG"), do: 5
  defp severity_number("INFO"), do: 9
  defp severity_number("WARN"), do: 13
  defp severity_number("ERROR"), do: 17
  defp severity_number("FATAL"), do: 21

  defp default_id do
    "elixir-#{System.unique_integer([:monotonic, :positive])}"
  end

  defp default_clock do
    DateTime.utc_now() |> DateTime.to_iso8601()
  end
end
